File: src/stroke.rs
use std::collections::{HashMap, HashSet};
use lazy_static::lazy_static;
use std::sync::Mutex;

lazy_static! {
    static ref STROKE_CONFIG: Mutex<Option<StrokeConfig>> = Mutex::new(None);
}

#[derive(Clone, Debug)]
pub struct StrokeConfig {
    pub keys: Vec<String>,
    pub implicit_hyphen_keys: HashSet<String>,
    pub number_key: Option<String>,
    pub numbers: HashMap<String, String>,
    pub undo_stroke_steno: String,
    // Derived
    pub key_order: HashMap<String, usize>,
}

pub fn setup_stroke(
    keys: Vec<String>,
    implicit_hyphen_keys: HashSet<String>,
    number_key: Option<String>,
    numbers: HashMap<String, String>,
    undo_stroke_steno: String,
) {
    let mut key_order = HashMap::new();
    for (i, key) in keys.iter().enumerate() {
        key_order.insert(key.clone(), i);
        if let Some(num) = numbers.get(key) {
            key_order.insert(num.clone(), i);
        }
    }

    let config = StrokeConfig {
        keys,
        implicit_hyphen_keys,
        number_key,
        numbers,
        undo_stroke_steno,
        key_order,
    };

    *STROKE_CONFIG.lock().unwrap() = Some(config);
}

pub fn get_stroke_config() -> StrokeConfig {
    STROKE_CONFIG
        .lock()
        .unwrap()
        .clone()
        .expect("Stroke system not initialized")
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Default)]
pub struct Stroke {
    value: u32,
}

impl Stroke {
    pub const fn new(value: u32) -> Self {
        Stroke { value }
    }

    pub fn from_steno(steno: &str) -> Self {
        let value = stroke_from_steno(steno);
        Stroke { value }
    }

    pub fn to_rtfcre(&self) -> String {
        stroke_to_steno(self.value)
    }

    pub fn is_correction(&self) -> bool {
        let cfg = get_stroke_config();
        self.value == Stroke::from_steno(&cfg.undo_stroke_steno).value
    }

    pub fn contains(&self, other: &Stroke) -> bool {
        (self.value & other.value) == other.value
    }

    pub fn subtract(&self, other: &Stroke) -> Stroke {
        Stroke { value: self.value & !other.value }
    }
}

fn stroke_from_steno(steno: &str) -> u32 {
    let cfg = get_stroke_config();
    if steno.is_empty() {
        return 0;
    }

    let mut keys = Vec::new();
    let mut remaining = steno;
    let mut has_number = false;

    if let Some(ref nk) = cfg.number_key {
        if remaining.starts_with('#') {
            has_number = true;
            remaining = &remaining[1..];
        }
    }

    let hyphen_index = remaining.find('-');
    let has_explicit_hyphen = hyphen_index.is_some();

    // Find implicit hyphen
    let mut implicit_hyphen_pos = None;
    if !has_explicit_hyphen {
        for (i, _) in remaining.char_indices() {
             let suffix = &remaining[i..];
             for impl_key in &cfg.implicit_hyphen_keys {
                 let key_char = impl_key.replace('-', "");
                 if suffix.starts_with(&key_char) {
                     implicit_hyphen_pos = Some(i);
                     break;
                 }
             }
             if implicit_hyphen_pos.is_some() { break; }
        }
    }

    let boundary_pos = if let Some(idx) = hyphen_index {
        idx
    } else {
        implicit_hyphen_pos.unwrap_or(remaining.len()) // Default to end if not found
    };

    // Left part
    let left_part = &remaining[0..boundary_pos];
    let mut left_remaining = left_part;

    while !left_remaining.is_empty() {
        let mut matched = false;
        for key in &cfg.keys {
            if key.starts_with('-') { continue; }
            let key_char = key.replace('-', "");

            if left_remaining.starts_with(&key_char) {
                 // Check number substitution
                 if has_number {
                     if let Some(num_val) = cfg.numbers.get(key) {
                         let num_char = num_val.replace('-', "");
                         if left_remaining.starts_with(&num_char) {
                             // This block logic from TS seems slightly different
                             // In TS: if hasNumber && numbers.has(key) && leftRemaining.startsWith(numChar) -> push(key)
                             // But wait, we are iterating keys, so we check if text matches key OR number char
                             // The TS logic iterates keys, then checks if text matches keyChar OR numChar
                         }
                     }
                 }

                 keys.push(key.clone());
                 left_remaining = &left_remaining[key_char.len()..];
                 matched = true;
                 break;
            }

            // Check number char directly
            if has_number {
                if let Some(num_val) = cfg.numbers.get(key) {
                    let num_char = num_val.replace('-', "");
                     if left_remaining.starts_with(&num_char) {
                        keys.push(key.clone());
                        left_remaining = &left_remaining[num_char.len()..];
                        matched = true;
                        break;
                     }
                }
            }
        }
        if !matched {
            // consume one char? In TS: leftRemaining.slice(1)
            let mut chars = left_remaining.chars();
            chars.next();
            left_remaining = chars.as_str();
        }
    }

    // Right part
    let right_part = if has_explicit_hyphen {
        &remaining[hyphen_index.unwrap() + 1..]
    } else if let Some(pos) = implicit_hyphen_pos {
        &remaining[pos..]
    } else {
        ""
    };

    let mut right_remaining = right_part;
    while !right_remaining.is_empty() {
        let mut matched = false;
        for key in &cfg.keys {
            let key_char = key.replace('-', "");
            if key.starts_with('-') || cfg.implicit_hyphen_keys.contains(key) {
                if right_remaining.starts_with(&key_char) {
                    if !keys.contains(key) {
                        keys.push(key.clone());
                    }
                    right_remaining = &right_remaining[key_char.len()..];
                    matched = true;
                    break;
                }
                 // Check number char directly
                if has_number {
                    if let Some(num_val) = cfg.numbers.get(key) {
                        let num_char = num_val.replace('-', "");
                         if right_remaining.starts_with(&num_char) {
                            if !keys.contains(key) {
                                keys.push(key.clone());
                            }
                            right_remaining = &right_remaining[num_char.len()..];
                            matched = true;
                            break;
                         }
                    }
                }
            }
        }
        if !matched {
             let mut chars = right_remaining.chars();
            chars.next();
            right_remaining = chars.as_str();
        }
    }

    if has_number {
        if let Some(nk) = &cfg.number_key {
            if !keys.contains(nk) {
                keys.push(nk.clone());
            }
        }
    }

    // Auto detect number key from content? TS does this.
    // Simplifying for now.

    stroke_from_keys(&keys)
}


fn stroke_from_keys(keys: &[String]) -> u32 {
    let cfg = get_stroke_config();
    let mut value = 0;
    for key in keys {
        if let Some(&idx) = cfg.key_order.get(key) {
            value |= 1 << idx;
        }
    }
    value
}

fn stroke_to_steno(value: u32) -> String {
    let cfg = get_stroke_config();
    let mut keys = Vec::new();
    for (i, key) in cfg.keys.iter().enumerate() {
        if (value & (1 << i)) != 0 {
            keys.push(key.clone());
        }
    }

    // Sort logic is implicit in iteration order of cfg.keys if they are in order

    if keys.is_empty() {
        return String::new();
    }

    let has_number_key = if let Some(nk) = &cfg.number_key {
        keys.contains(nk)
    } else {
        false
    };

    let mut result = String::new();
    let mut need_hyphen = true;
    let mut past_implicit = false;

    for key in &keys {
        if Some(key) == cfg.number_key.as_ref() {
            continue;
        }

        let key_char = key.replace('-', "");

        if cfg.implicit_hyphen_keys.contains(key) {
            past_implicit = true;
            need_hyphen = false;
        }

        if key.starts_with('-') && need_hyphen && !past_implicit {
            result.push('-');
            need_hyphen = false;
        }

        if has_number_key {
            if let Some(num) = cfg.numbers.get(key) {
                result.push_str(&num.replace('-', ""));
            } else {
                result.push_str(&key_char);
            }
        } else {
             result.push_str(&key_char);
        }
    }

    if has_number_key {
        // Simple check if result has numbers
        let has_num_char = cfg.numbers.values().any(|v| result.contains(&v.replace('-', "")));
        if !has_num_char {
             result.insert(0, '#');
        }
    }

    result
}

pub fn normalize_steno(steno: &str) -> Vec<String> {
    steno.split('/')
        .map(|s| Stroke::from_steno(s).to_rtfcre())
        .collect()
}

File: src/stroke.ts
/**
 * Stroke - A TypeScript implementation of steno stroke handling
 * 
 * This module provides stroke parsing, normalization, and manipulation
 * for stenography. It's a pure TypeScript port of the plover_stroke
 * C extension functionality.
 */

export interface StrokeConfig {
  keys: readonly string[];
  implicitHyphenKeys: Set<string>;
  numberKey: string | null;
  numbers: Map<string, string>;
  feralNumberKey: boolean;
  undoStrokeSteno: string;
}

// Global configuration - set by system.setup()
let config: StrokeConfig | null = null;
let keyOrder: Map<string, number> = new Map();
let reverseNumbers: Map<string, string> = new Map();
let implicitHyphens: Set<string> = new Set();

/**
 * Setup the stroke system with the given configuration
 */
export function setupStroke(cfg: StrokeConfig): void {
  config = cfg;
  
  // Build key order map
  keyOrder = new Map();
  for (let i = 0; i < cfg.keys.length; i++) {
    keyOrder.set(cfg.keys[i], i);
    const numberKey = cfg.numbers.get(cfg.keys[i]);
    if (numberKey) {
      keyOrder.set(numberKey, i);
    }
  }
  
  // Build reverse numbers map
  reverseNumbers = new Map();
  for (const [key, num] of cfg.numbers) {
    reverseNumbers.set(num, key);
  }
  
  // Build implicit hyphens set (without the hyphen)
  implicitHyphens = new Set();
  for (const key of cfg.implicitHyphenKeys) {
    implicitHyphens.add(key.replace('-', ''));
  }
}

/**
 * Get the current stroke configuration
 */
export function getStrokeConfig(): StrokeConfig {
  if (!config) {
    throw new Error('Stroke system not initialized. Call setupStroke() first.');
  }
  return config;
}

/**
 * Represents a stenotype stroke as an integer bitmask
 */
export class Stroke {
  private _value: number;
  private _rtfcre: string | null = null;
  private _keys: string[] | null = null;

  constructor(value: number | string | string[]) {
    if (typeof value === 'number') {
      this._value = value;
    } else if (typeof value === 'string') {
      this._value = Stroke.fromSteno(value)._value;
    } else if (Array.isArray(value)) {
      this._value = Stroke.fromKeys(value)._value;
    } else {
      throw new Error('Invalid stroke value');
    }
  }

  /**
   * Get the integer value of the stroke
   */
  get value(): number {
    return this._value;
  }

  /**
   * Get the RTFCRE string representation
   */
  get rtfcre(): string {
    if (this._rtfcre === null) {
      this._rtfcre = strokeToSteno(this._value);
    }
    return this._rtfcre;
  }

  /**
   * Get the steno keys in this stroke
   */
  get stenoKeys(): string[] {
    if (this._keys === null) {
      this._keys = strokeToKeys(this._value);
    }
    return this._keys;
  }

  /**
   * Check if this is an undo/correction stroke
   */
  get isCorrection(): boolean {
    const cfg = getStrokeConfig();
    const undoStroke = Stroke.fromSteno(cfg.undoStrokeSteno);
    return this._value === undoStroke._value;
  }

  /**
   * Get the number of keys in this stroke
   */
  get length(): number {
    return this.stenoKeys.length;
  }

  /**
   * Create a stroke from steno notation
   */
  static fromSteno(steno: string): Stroke {
    return new Stroke(strokeFromSteno(steno));
  }

  /**
   * Create a stroke from a list of keys
   */
  static fromKeys(keys: string[]): Stroke {
    return new Stroke(strokeFromKeys(keys));
  }

  /**
   * Create a stroke from an integer
   */
  static fromInteger(value: number): Stroke {
    return new Stroke(value);
  }

  /**
   * Normalize a single stroke string
   */
  static normalizeStroke(steno: string, strict = true): string {
    try {
      return Stroke.fromSteno(steno).rtfcre;
    } catch (e) {
      if (strict) throw e;
      return steno;
    }
  }

  /**
   * Normalize a steno outline (potentially multi-stroke)
   */
  static normalizeSteno(steno: string, strict = true): string[] {
    try {
      return steno.split('/').map(s => Stroke.normalizeStroke(s, strict));
    } catch (e) {
      if (strict) throw e;
      return steno.split('/');
    }
  }

  /**
   * Get a sort key for a steno outline
   */
  static stenoToSortKey(steno: string, strict = true): string {
    try {
      const strokes = steno.split('/').map(s => Stroke.fromSteno(s));
      // Create a sortable string based on stroke values
      return strokes.map(s => s._value.toString(16).padStart(8, '0')).join('');
    } catch (e) {
      if (strict) throw e;
      // Fallback: prefix with null bytes to sort after valid strokes
      return '\x00\x00' + steno;
    }
  }

  /**
   * The prefix stroke (empty stroke)
   */
  static get PREFIX_STROKE(): Stroke {
    return new Stroke(0);
  }

  /**
   * The undo stroke
   */
  static get UNDO_STROKE(): Stroke {
    const cfg = getStrokeConfig();
    return Stroke.fromSteno(cfg.undoStrokeSteno);
  }

  /**
   * Check if this stroke contains another stroke (bitwise)
   */
  contains(other: Stroke): boolean {
    return (this._value & other._value) === other._value;
  }

  /**
   * Subtract another stroke (bitwise)
   */
  subtract(other: Stroke): Stroke {
    return new Stroke(this._value & ~other._value);
  }

  /**
   * Combine with another stroke (bitwise OR)
   */
  combine(other: Stroke): Stroke {
    return new Stroke(this._value | other._value);
  }

  /**
   * Check equality
   */
  equals(other: Stroke): boolean {
    return this._value === other._value;
  }

  /**
   * Get iterator over keys
   */
  keys(): string[] {
    return this.stenoKeys;
  }

  toString(): string {
    const prefix = this.isCorrection ? '*' : '';
    return `${prefix}Stroke(${this.rtfcre} : ${this.stenoKeys.join(', ')})`;
  }
}

// ============================================================================
// Internal Helper Functions
// ============================================================================

/**
 * Parse a steno string into a bitmask
 */
function strokeFromSteno(steno: string): number {
  const cfg = getStrokeConfig();
  
  if (!steno) {
    return 0;
  }

  const keys: string[] = [];
  let remaining = steno;
  
  // Handle number key prefix
  // Numeric aliases are themselves an RTFCRE representation of the number
  // bar.  The serializer normally omits `#` when at least one alias is
  // present, so detect them before parsing the individual keys.
  let hasNumber = cfg.numberKey !== null && [...cfg.numbers.values()].some(
    number => steno.includes(number.replace('-', ''))
  );
  if (cfg.numberKey) {
    const numberKey = cfg.numberKey.replace('-', '');
    const firstNumberKeyIndex = remaining.indexOf(numberKey);
    if (firstNumberKeyIndex !== -1) {
      const secondNumberKeyIndex = remaining.indexOf(
        numberKey,
        firstNumberKeyIndex + numberKey.length,
      );
      if (secondNumberKeyIndex !== -1) {
        throw new Error(`Duplicate number key: ${steno}`);
      }
      if (firstNumberKeyIndex !== 0 && !cfg.feralNumberKey) {
        throw new Error(`Number key must be leading: ${steno}`);
      }
      hasNumber = true;
      remaining =
        remaining.slice(0, firstNumberKeyIndex) +
        remaining.slice(firstNumberKeyIndex + numberKey.length);
    }
  }
  
  // Check for explicit hyphen to determine left/right boundary
  const hyphenIndex = remaining.indexOf('-');
  const hasExplicitHyphen = hyphenIndex !== -1;
  
  // Find implicit hyphen position if no explicit hyphen
  let implicitHyphenPos = -1;
  if (!hasExplicitHyphen) {
    for (let i = 0; i < remaining.length; i++) {
      for (const implKey of cfg.implicitHyphenKeys) {
        const keyChar = implKey.replace('-', '');
        if (remaining.slice(i).startsWith(keyChar)) {
          implicitHyphenPos = i;
          break;
        }
      }
      if (implicitHyphenPos !== -1) break;

      // A numeric alias for a vowel or right-hand key establishes the same
      // boundary as the underlying steno key.
      if (hasNumber) {
        for (const [key, number] of cfg.numbers) {
          if ((cfg.implicitHyphenKeys.has(key) || key.startsWith('-')) &&
              remaining.slice(i).startsWith(number.replace('-', ''))) {
            implicitHyphenPos = i;
            break;
          }
        }
      }
      if (implicitHyphenPos !== -1) break;
    }
  }
  
  // Determine the boundary position
  const boundaryPos = hasExplicitHyphen ? hyphenIndex : implicitHyphenPos;
  
  // Build list of left-hand keys (before boundary)
  const leftPart = boundaryPos === -1 ? remaining : remaining.slice(0, boundaryPos);
  
  // Build list of right-hand keys (after boundary)
  let rightPart = '';
  if (hasExplicitHyphen) {
    rightPart = remaining.slice(hyphenIndex + 1);
  } else if (implicitHyphenPos !== -1) {
    rightPart = remaining.slice(implicitHyphenPos);
  }
  
  // Parse left-hand keys
  let leftRemaining = leftPart;
  while (leftRemaining.length > 0) {
    let matched = false;
    
    // Try to match left-hand keys (keys ending with - or without hyphen that are left)
    for (const key of cfg.keys) {
      // Skip right-hand keys
      if (key.startsWith('-')) continue;
      
      const keyChar = key.replace('-', '');
      if (leftRemaining.startsWith(keyChar)) {
        // Check for number substitution
        if (hasNumber && cfg.numbers.has(key)) {
          const numChar = cfg.numbers.get(key)!.replace('-', '');
          if (leftRemaining.startsWith(numChar)) {
            keys.push(key);
            leftRemaining = leftRemaining.slice(numChar.length);
            matched = true;
            break;
          }
        }
        
        keys.push(key);
        leftRemaining = leftRemaining.slice(keyChar.length);
        matched = true;
        break;
      }
      
      // Also check for number character
      if (hasNumber && cfg.numbers.has(key)) {
        const numChar = cfg.numbers.get(key)!.replace('-', '');
        if (leftRemaining.startsWith(numChar)) {
          keys.push(key);
          leftRemaining = leftRemaining.slice(numChar.length);
          matched = true;
          break;
        }
      }
    }
    
    if (!matched) {
      // Skip unknown character
      leftRemaining = leftRemaining.slice(1);
    }
  }
  
  // Parse right-hand keys (including implicit hyphen keys)
  let rightRemaining = rightPart;
  while (rightRemaining.length > 0) {
    let matched = false;
    
    for (const key of cfg.keys) {
      const keyChar = key.replace('-', '');
      
      // Match implicit hyphen keys or right-hand keys
      if (cfg.implicitHyphenKeys.has(key) || key.startsWith('-')) {
        if (rightRemaining.startsWith(keyChar)) {
          // Check for number substitution
          if (hasNumber && cfg.numbers.has(key)) {
            const numChar = cfg.numbers.get(key)!.replace('-', '');
            if (rightRemaining.startsWith(numChar)) {
              keys.push(key);
              rightRemaining = rightRemaining.slice(numChar.length);
              matched = true;
              break;
            }
          }
          
          if (!keys.includes(key)) {
            keys.push(key);
          }
          rightRemaining = rightRemaining.slice(keyChar.length);
          matched = true;
          break;
        }
        
        // Also check for number character
        if (hasNumber && cfg.numbers.has(key)) {
          const numChar = cfg.numbers.get(key)!.replace('-', '');
          if (rightRemaining.startsWith(numChar)) {
            keys.push(key);
            rightRemaining = rightRemaining.slice(numChar.length);
            matched = true;
            break;
          }
        }
      }
    }
    
    if (!matched) {
      // Skip unknown character
      rightRemaining = rightRemaining.slice(1);
    }
  }
  
  // Add number key if needed
  if (hasNumber && cfg.numberKey && !keys.includes(cfg.numberKey)) {
    keys.push(cfg.numberKey);
  }
  
  // Also detect numbers from the content
  if (!hasNumber && cfg.numberKey) {
    for (const numChar of cfg.numbers.values()) {
      if (steno.includes(numChar.replace('-', ''))) {
        if (!keys.includes(cfg.numberKey)) {
          keys.push(cfg.numberKey);
        }
        break;
      }
    }
  }

  return strokeFromKeys(keys);
}

/**
 * Convert a list of keys to a bitmask
 */
function strokeFromKeys(keys: string[]): number {
  const cfg = getStrokeConfig();
  let value = 0;
  
  for (const key of keys) {
    const order = keyOrder.get(key);
    if (order !== undefined) {
      value |= (1 << order);
    }
  }
  
  return value;
}

/**
 * Convert a bitmask to a list of keys
 */
function strokeToKeys(value: number): string[] {
  const cfg = getStrokeConfig();
  const keys: string[] = [];
  
  for (let i = 0; i < cfg.keys.length; i++) {
    if (value & (1 << i)) {
      keys.push(cfg.keys[i]);
    }
  }
  
  return keys;
}

/**
 * Convert a bitmask to steno notation
 */
function strokeToSteno(value: number): string {
  const cfg = getStrokeConfig();
  const keys = strokeToKeys(value);
  
  if (keys.length === 0) {
    return '';
  }

  // Sort keys by their order
  keys.sort((a, b) => (keyOrder.get(a) ?? 0) - (keyOrder.get(b) ?? 0));

  // Check if we have the number key
  const hasNumberKey = cfg.numberKey !== null && keys.includes(cfg.numberKey);
  
  // Build the steno string
  let result = '';
  let needHyphen = true;
  let pastImplicit = false;

  for (const key of keys) {
    // Skip the number key for now
    if (key === cfg.numberKey) continue;

    const keyChar = key.replace('-', '');
    
    // Check if this is an implicit hyphen key
    if (cfg.implicitHyphenKeys.has(key)) {
      pastImplicit = true;
      needHyphen = false;
    }
    
    // Check if this is a right-hand key (starts with -)
    if (key.startsWith('-') && needHyphen && !pastImplicit) {
      result += '-';
      needHyphen = false;
    }

    // Convert to number if applicable
    if (hasNumberKey && cfg.numbers.has(key)) {
      const numKey = cfg.numbers.get(key)!;
      result += numKey.replace('-', '');
    } else {
      result += keyChar;
    }
  }

  // Add number key prefix if needed and not already represented
  if (hasNumberKey) {
    // Check if any number character is in the result
    let hasNumChar = false;
    for (const numKey of cfg.numbers.values()) {
      if (result.includes(numKey.replace('-', ''))) {
        hasNumChar = true;
        break;
      }
    }
    if (!hasNumChar) {
      result = '#' + result;
    }
  }

  return result;
}

// Export helper functions
export const normalizeStroke = Stroke.normalizeStroke;
export const normalizeSteno = Stroke.normalizeSteno;
export const stenoToSortKey = Stroke.stenoToSortKey;

/**
 * Sort steno strokes by fewest strokes, then fewest keys
 */
export function sortStenoStrokes(strokesList: string[][]): string[][] {
  return [...strokesList].sort((a, b) => {
    // First by number of strokes
    if (a.length !== b.length) {
      return a.length - b.length;
    }
    // Then by total number of keys
    const aKeys = a.reduce((sum, s) => sum + Stroke.fromSteno(s).length, 0);
    const bKeys = b.reduce((sum, s) => sum + Stroke.fromSteno(s).length, 0);
    return aKeys - bKeys;
  });
}
